File: ex01/Vector.py
class Vector:
    values = []

    def __init__(self, values):
        self.values = values

    def add(self, v2):
        self.values = [self.values[i] + v2.values[i]
                       for i in range(len(self.values))]
        return Vector(self.values)

    def sub(self, v2):
        self.values = [self.values[i] - v2.values[i]
                       for i in range(len(self.values))]
        return Vector(self.values)

    def scl(self, scaler):
        self.values = [self.values[i] *
                       scaler for i in range(len(self.values))]
        return Vector(self.values)

    def __str__(self):
        return str(self.values)

File: ex03/vector.cpp
#include "vector.hpp"

Vector::Vector(const std::vector<float>& vals) : values(vals) {}

Vector Vector::add(const Vector& v2) {
    std::vector<float> result;
    for (size_t i = 0; i < values.size(); i++) {
        result.push_back(values[i] + v2.values[i]);
    }
    return Vector(result);
}

Vector Vector::sub(const Vector& v2) {
    std::vector<float> result;
    for (size_t i = 0; i < values.size(); i++) {
        result.push_back(values[i] - v2.values[i]);
    }
    return Vector(result);
}

Vector Vector::scl(float scaler) {
    std::vector<float> result;
    for (size_t i = 0; i < values.size(); i++) {
        result.push_back(values[i] * scaler);
    }
    return Vector(result);
}

float Vector::dot(const Vector& v2) {
   int i = this->values.size();
    float sum = 0.0f;
   for (int i = 0; i < this->values.size() && i < v2.values.size(); i++){
        sum = fma(this->values[i], v2.values[i], sum);
   }
   return sum;
}

std::string Vector::toString() const {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < values.size(); ++i) {
        os << values[i];
        if (i != values.size() - 1) os << ", ";
    }
    os << "]";
    return os.str();
}

std::ostream& operator<<(std::ostream& os, const Vector& v) {
    os << v.toString();
    return os;
}
